packages/camera/camera/ios/Classes/FLTThreadSafeEventChannel.h

```diff
@@ -0,0 +1,29 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#import <Flutter/Flutter.h>
+
+NS_ASSUME_NONNULL_BEGIN
+
+/**
+ * Wrapper for FlutterEventChannel that always sends events on the main thread
```

Comment:
Yes these objects are thread safe, though it'd be a bit odd to me to add a duplicate comment saying a "FLTThreadSafeFoo" is thread safe. The class name commented itself.

Reply:
But the description just says that they always call things on the main thread. The fact that its purpose is to make *other code* threadsafe leaves a lot of ambiguity here.

If you think the current naming is completely self-explanatory: what would you call a class whose job is to always funnel `EventChannel` calls to the main thread, but was itself thread-hostile for some reason?

Reply:
I'm thinking `A thread safe wrapper for FlutterEventChannel that can be called from any threads`. The "dispatching to main thread" part is more like implementation details that the consumer does not need to know.  What do you think?

Reply:
I don't think it's really an implementation detail; it's the reason for the existence of the class. If at some point the need to do that were to change (e.g., exactly what is happening with `FlutterResult` right now), the whole class could be removed.

Reply:
To clarify, I think "dispatching to main" is the implementation detail that *the caller of the wrapper* should not care about. The caller only cares the wrapper API is thread-safe, and does not care about the fact that the underlying engine api must happen on main and the thread-safety is achieved by dispatching to main. 

In the future, when the original API itself becomes thread-safe, we can remove this wrapper. In that world, how the original API achieves thread-safety is also implementation detail that the caller should not care about. It only cares that the new API can be called from any thread. As to how this is achieved? By dispatching to a the same serial queue like what we do here? or by using other mechanisms like atomic CAS? or by other means? it doesn't matter.

Reply:
In the current world, the consumer of the engine API is the wrapper *implementation*, and the consumer of the wrapper API is in `CameraPlugin.m`.

Reply:
> To clarify, I think "dispatching to main" is the implementation detail that _the consumer of the wrapper_ should not care about.

I understand who we're speaking in reference to when we say that something is or is not an implementation detail.

> The consumer only cares the wrapper API is thread-safe, and does not care about the fact that the underlying engine api must happen on main and the thread-safety is achieved by dispatching to main.

I strongly disagree. The consumer is a plugin. A plugin is written against the engine APIs, and is responsible for ensuring that the APIs are used correctly, including calling main-thread-APIs on the main thread. This class is a tool for doing that, it's not a complete abstraction that eliminates the plugin code's knowledge of the Flutter engine.

**How** it dispatches to the main thread is an implementation detail. The fact that it does so is integral to the API contract of the class itself.

> In the future, when the original API itself becomes thread-safe, we can remove this wrapper. In that world, how the original API achieves thread-safety is also implementation detail that the consumer of the API should not care about.

The difference is that the engine is a complete API surface that sets threading requirements. This is a class that is **not part of that API** whose purpose is to make it easier for code to meet certain threading restrictions on an object **from that API surface**.

Reply:
I think there are 2 parts discussed here - the wrapper, and the consumer of the wrapper. Only the wrapper's implementation cares about the knowledge that engine API must be run on main. This knowledge is abstracted away from the consumer of the wrapper. 

Yes the wrapper is not part of the engine, but it can still encapsulate implementation details and hide it from its consumers. Having the wrapper to be inside the same plugin directory as its consumers does not mean its consumers must know the internal of this wrapper class.

Reply:
> I strongly disagree. The consumer is a plugin. 

Looks like miscommunication here. To clarify, the "consumer" quoted here is to refer to the "consumer of the wrapper" in the previous sentence, and not the "consumer of the engine API". 

So let me emphasize it again because it's very important: 
- The consumer of the engine API is the plugin code, or more precisely, the wrapper part of the plugin code.
- The consumer of the wrapper API is the `CameraPlugin.m` part of the plugin code. 

The wrapper and the consumer of the wrapper API are both under the same plugin directory, but it does not mean that the wrapper's implementation detail has to be exposed to the consumer of the wrapper. The wrapper encapsulated its internal implementation details of how to achieve the thread-safety.

Reply:
> Looks like miscommunication here. To clarify, the "consumer" quoted here is to refer to the "consumer of the wrapper" in the previous sentence, and not the "consumer of the engine API".

I understand how this code works, and who the consumers of each API are. It would be helpful if you could stop assuming that any time I disagree with your conclusion I must not understand the code.

> The wrapper and the consumer of the wrapper API are both under the same plugin directory, but it does not mean that the wrapper's implementation detail has to be exposed to the consumer of the wrapper.

I understand the concept that directories and API surfaces are not the same thing.

> The wrapper encapsulated its internal implementation details of how to achieve the thread-safety.

I have already said I agree with this statement. Whether it uses GCD, operation queues, low-level posix primitives, etc., is an implementation detail that is, correctly, encapsulated.

> * Does the `CameraPlugin.m` part of the plugin know that engine API has to be run on the main thread?
>   No, `CameraPlugin.m` only cares that the wrapper is thread safe, so can be called from any threads.

I don't understand the reason for the assertion that the code that **creates the event channel** has no responsibility for knowing anything about the threading model of the object that it is creating and handing off to an object that is explicitly for multi-threaded use. And I don't agree with it for reasons that I've explained.

Specifically, I disagree with this characterization of the code:

> So let me emphasize it again because it's very important:
>
> * The consumer of the engine API is the plugin code, or more precisely, the wrapper part of the plugin code.

`CameraPlugin.m` is a direct consumer of engine APIs, including a direct consumer of `EventChannel` (when it creates it). It is not a consumer of the the specific calls that are wrapped, that's true, but it is deeply, integrally intertwined with the engine API.

Here's a plausible hypothetical for you: imagine that at some point `EventChannel` were extended to have an `eventChannelWithName:binaryMessenger:queue:`, whose semantics were that it was still not thread-safe, but that when initialized that way all calls to it would need to be made from that queue rather than the main queue. This first line of this code **in `CameraPlugin.m`**:
```
    FlutterEventChannel *eventChannel =
        [FlutterEventChannel eventChannelWithName:@"plugins.flutter.io/camera/imageStream"
                                  binaryMessenger:messenger];
    FLTThreadSafeEventChannel *threadSafeEventChannel =
        [[FLTThreadSafeEventChannel alloc] initWithEventChannel:eventChannel];
```

is changed to:

```
    FlutterEventChannel *eventChannel =
        [FlutterEventChannel eventChannelWithName:@"plugins.flutter.io/camera/imageStream"
                                  binaryMessenger:messenger
                                            queue:someOtherQueue];
```

The threading of `FLTThreadSafeEventChannel` is now completely broken by a change in its consumer. If this were a true abstraction of the threading model of the Flutter engine, that would be impossible.

It would be possible to build a complete abstraction layer between all the real logic of the camera plugin and the Flutter engine, but these utility classes **do not do that**.

> Can we break this encapsulation? Well I don't recommend it, but sure.

My point is that I do not agree that this level of encapsulation exists here in the first place. `CameraPlugin.m` is not a higher level of abstraction that is blissfully unaware that it is interacting with the Flutter engine, such that all of the Flutter engine's API requirements are irrelevant to it. It is the central interface with the engine APIs in general.

If I were new to this plugin but familiar with plugin threading, and reading `CameraPlugin.m` to understand how it works (e.g., to debug an issue that could be related to threading), and I saw that a class I know is not safe to use from other threads was being handed to another class that is clearly intended for use from other threads, it would be a red flag. I would then go look at that class's header. If it said that it guarantees that the object is always called from the main thread, then I would immediately understand what it was doing and go back to the main logic. If it did not, then I would need to take a detour into reading the implementation of that class to see if the plugin had been correctly designed in the first place.

> If you have any thoughts on what that line of comment should be, I'm more than happy to update them.

Most of this thread started from your question here:

> The "dispatching to main thread" part is more like implementation details that the consumer does not need to know. What do you think?

I've said what I think: that we shouldn't remove that part, because I don't think its an implementation detail that is irrelevant to the consuming code. But you don't seem to be happy to make that change (or rather, not make the change of removing it) as evidenced by the length of this thread.

Reply:
> But you don't seem to be happy to make that change (or rather, not make the change of removing it) as evidenced by the length of this thread.

That's not the case at all. I am **extremely happy** to update the comment. It's just I wasn't clear what you wanted. I will take a look and digest your explanation above tomorrow. But just to avoid further churn, I will update it to - `A thread safe wrapper for FlutterEventChannel that can be called from any threads, by dispatching the underlying engine API call to the main thread. `.

Reply:
Just sent a meeting invite for a quick chat.

Reply:
Discussed offline. The key thing I overlooked is that, the `CameraPlugin.m` **is** a direct consumer of the engine API (e.g. creation of the thread-unsafe type to be wrapped). I was imagining a world where the `CameraPlugin.m` only interact with the wrapper, and the wrapper interact with the engine API. But that's not the case.